users/userRouter.js

```diff
@@ -1,47 +1,154 @@
+    .catch(error => {
+      console.log(error);
+      res.status(500).json({
+        message: "error retrieving the users"
+      })
+    })
 });
 
-router.get('/:id', (req, res) => {
+router.get('/:id', validateUserId, (req, res) => {
   // do your magic!
+  res.status(200).json(req.user)
+  //with validation baked in:
+  // Users.getById(req.params.id)
+  //   .then( user => {
+  //     user ? res.status(200).json(user)
+  //     : res.status(404).json({ message: "User not found" })
+  //   })
 });
 
-router.get('/:id/posts', (req, res) => {
+router.get('/:id/posts', validateUserId, (req, res) => {
   // do your magic!
+  Users.getUserPosts(req.params.id)
+    .then(posts => {
+      res.status(200).json(posts)
+    })
+    .catch(error => {
+      res.status(500).json({ message: "Posts not found" })
+    })
 });
 
 router.delete('/:id', (req, res) => {
   // do your magic!
+  Users.remove(req.params.id)
+    .then(count => {
+      if (count > 0) {
+        res.status(200).json({ message: "the user been erased from history"})
+      } else {
+        res.status(404).json({message: "the user couldn't be found"})
+      }
+    })
+    .catch(error => {
+      console.log(error)
+      res.status(500).json({
+        message: "error removing user"
+      })
+    })
 });
 
-router.put('/:id', (req, res) => {
+router.put('/:id', validateUserId, (req, res) => {
   // do your magic!
+  Users.update(req.params.id, req.body)
+  .then(count => {
+    if (count > 0) {
+      res.status(200).json({ message: "the user been updated"})
+    } else {
+      res.status(404).json({message: "the user couldn't be updated"})
+    }
+  })
+  .catch(error => {
+    console.log(error)
+    res.status(500).json({
+      message: "error updating user"
+    })
+  })
 });
 
 //custom middleware
 
 function validateUserId(req, res, next) {
   // do your magic!
+  Users.getById(req.params.id)
+    .then( user => {
+      if(user) {
+        req.user = user;
+        next();
+      } else {
+        res.status(404).json({ message: "User not found (Stopped by validateUserId)" });
+      }
+    })
+    .catch(error => {
+      res.status(400).json({ message: "invalid user id (Stopped by validateUserId)" })
+    });
 }
 
 function validateUser(req, res, next) {
   // do your magic!
+  console.log(req.body)
+  if (req.body) {
+    req.body.name ? next()
+    : res.status(400).json({ message: "missing required name field" })
+  } else {
+    //###Q: in practice I can't get hit this error, even when setting postman to no body. Why?
+    res.status(400).json({ message: "missing user data" })
+  }
 }
 
 function validatePost(req, res, next) {
   // do your magic!
+  console.log(req.body)
+  if (req.body) {
+    req.body.text ? next()
+    : res.status(400).json({ message: "missing required text field" })
+  } else {
+    //###Q: in practice I can't get hit this error, even when setting postman to no body. Why?
+    res.status(400).json({ message: "missing user data" })
+  }
 }
```

Comment:
`if( req.body)` was always true, even when I intentionally used postman with no body. Why so? I couldn't get it to trigger the else statement.

Reply:
![image](https://user-images.githubusercontent.com/980338/94771568-602f5c80-036c-11eb-93ad-e580a062d065.png)

Reply:
```js
function validatePost(req, res, next) {
  if (!req.body.text) {
    return res.status(400).json({ message: "missing required text field" })
  }

  if(!Object.keys(req.body).length) {
    return res.status(400).json({ message: "missing user data" })
  }

  return next()
}
```